Added options to make PR compare against base branch and then output result

pksunkara opened the pull request:
https://github.com/clap-rs/clap/pull/1724 is the PR you can see.

As you can see there, both the options are working.

rhysd commented:
Thank you for adding this. Let me review this PR this weekend.

pksunkara commented:
Hey, any update on this?

rhysd commented:
I'm sorry for the delay. I'm now focusing on another private project and could not review this last weekend. Let me check now.

rhysd commented on src/extract.ts:
```diff
@@ -363,7 +363,7 @@ function extractCatch2Result(output: string): BenchmarkResult[] {
         }
 
-        const range = '+/- ' + stdDev[1].trim();
+        const range = '± ' + stdDev[1].trim();
```
Would you tell me why this change is necessary? I'm not against `±`, but I wonder what is the better than `+/-`.

pksunkara replied:
We have a lot of benchmarks, which means the names of those benchmarks are a bit longer. This reduces the width of the table that's commented on the commit which means it looks nicer.

rhysd replied:
OK, that makes sense.

rhysd commented on src/write.ts:
```diff
@@ -117,7 +117,58 @@ function getCurrentRepo() {
 
 function floatStr(n: number) {
-    return Number.isInteger(n) ? n.toFixed(1) : n.toString();
+    return Number.isInteger(n) ? n.toFixed(0) : n.toFixed(2);
```
Why is this changed? I think the `n.toFixed(2)` is not acceptable. For example, when the value is much smaller than zero, the value is truncated to zero.

```
> 0.00001.toFixed(2)
'0.00'
```

Please keep `toString()` here.

pksunkara replied:
We are using this function only for thresholds which are never near to `0`. Having it `toString()` printed out more than 10 character string which I said before doesn't make the comment look nice because too much width on the table.

rhysd replied:
This action is not only for your case. Unless you ensure that all benchmark tools' outputs, please use `toString` conservatively.

pksunkara replied:
Can I add an option for this then?

rhysd replied:
How about make the logic cleverer to fit your case?

For example:

```javascript
if (Number.isInteger(n)) {
    return n.toString();
}

if (n > 0) {
    return n.toFixed(2);
}

return n.toString(); // Fallback
```

Here when `n` is larger than zero, it means that it's big enough and we can truncate the number. So we can use `toFixed(2)`. Otherwise we faillback into  `toString`.

Introducing new option also introduces more complexity. So I'm trying not to add new option as much as possible.

BTW, what is the 'ugly' case you're facing (just curious)?

pksunkara replied:
You mean `n > 1` in the if loop condition, right?

Check https://github.com/clap-rs/clap/pull/1724#commitcomment-37649200. That much info is not needed for the ratio.

So, I updated the PR with your proposed change.

rhysd replied:
Yes, you're correct. Thanks, I'll check it.

pksunkara replied:
So, I realized the condition should probably be `n > 0.1`, not `n > 1`. Because if n is 0.98499484849, then the whole thing will be printed.

rhysd replied:
ah, I see. I'll fix up before release

rhysd commented on src/write.ts:
```diff
@@ -117,7 +117,58 @@ function getCurrentRepo() {
+    // eslint-disable-next-line @typescript-eslint/camelcase
+    const repoUrl = repo.html_url ?? '';
+    const actionUrl = repoUrl + '/actions?query=workflow%3A' + encodeURIComponent(github.context.workflow);
+    core.debug(`Action URL: ${actionUrl}`);
```
Would you remove this debug print? We can check the `actionUrl` value in comment. So printing it in job is not so useful.

pksunkara commented on README.md:
```diff
@@ -368,6 +368,23 @@ Please see the 'Commit comment' section for more details.
+- Default: `true`
+
+If it is set to `true`, any commits that are part of a PR will not save the current benchmark.
+You can use this option to set up your action to compare the benchmarks between PR and base branch.
```
This is mostly about external data json. `auto-push` is only used for storing in pages. I need to compare PR's benchmark data with master's benchmark data. Currently, the logic is too much interwined with saving benchmarks because that's how we need to get previous data. I will look into this. And I will unify the option with `auto-push` and name it as `save` while maintaining backwards compatibility. What do you think?

pksunkara replied:
The logic is too much interwined. So, I just used the `auto-push` setting for saving to external data file also

rhysd replied:
OK, would you make the `save-on-pr` not only for PR? I feel the feature is too specific if it is only for PR. Instead, would you make a `save-data-file` boolean option which controls whether saving the results to file or not. It is more general because some people may want to be notified by comment every commit on `push` event. In your use case, `if` is available to control the option in step as above example.

pksunkara commented:
I removed the `saveOnPr` option and used `autoPush`. But it would break backward compatibility for people who don't have `autoPush` set to `true` and relying that their benchmarks are stored to external data files.

rhysd commented on src/write.ts:
```diff
@@ -401,10 +454,15 @@ async function writeBenchmarkToExternalJson(
+    if (!autoPush) {
+        core.debug('Skipping storing benchmarks in external data file');
+        return prevBench;
+    }
```
Please avoid this guard. This breaks CI and does not behave as description of `auto-push`. `auto-push` controls whether pushing commit to remote or not. It means that data should be saved to file and commit needs be created. `auto-push` option is for those who want to push the change to remote by themselves.

rhysd commented on test/write.ts:
```diff
@@ -186,8 +186,9 @@ describe('writeBenchmark()', function() {
             benchmarkDataDirPath: 'dummy', // Should not affect
             githubToken: undefined,
-            autoPush: false,
+            autoPush: true,
```
Why is this changed?

rhysd commented:
Thanks for the responses. I also left a comment in outdated conversation. https://github.com/rhysd/github-action-benchmark/pull/17#discussion_r390979472

pksunkara commented:
@rhysd I have addressed your review. The only thing left to decide is https://github.com/rhysd/github-action-benchmark/pull/17#discussion_r391187071. I really want some customization here because it looks really ugly otherwise.

rhysd commented:
@pksunkara I merged this. Thank you for your contribution. I'll make a new release tomorrow.

rhysd commented:
I'm sorry for the delay. I released new version v1.8.0 including this patch.
